### PR TITLE
2.5 database migration and hohonu ingestion class

### PR DESCRIPTION
**Pull request by abeasley1722**
Changes:
2.5 Database Migration
Hohonu Ingestion Class
Magnolia Beach ChartJson

Overall, just test that the database migration works as intended(adding dataSource and dataMapping) and that the hohonu test works. I will send the .env needed in email.

### CODE REVIEW

File: src/DataIngestion/DI_Classes/HOHONU.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
+#!/usr/local/bin/python
+#NDBC.py
```
Comment (matdenkas) [inline src/DataIngestion/DI_Classes/HOHONU.py:2]: Super nit picky NDBC -> HOHONU

File: src/DataIngestion/DI_Classes/HOHONU.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        """
+        # Convert datetime to ISO 8601 string format
+        dt_str = dt.strftime('%Y-%m-%dT%H:%M:%SZ')
+        
+        # URL-encode the string representation
+        dt_quoted = quote(dt_str)
+        
+        return dt_quoted
+    
+    def mm_to_m(self, mm):
```
Comment (matdenkas) [inline src/DataIngestion/DI_Classes/HOHONU.py:199]: Method should be private. Again super nit picky feel free to ignore

File: src/Frontend/ChartJson/Magnolia_Bay.json [hunk 3]
```diff
@@ -0,0 +1,123 @@
+{
```
Comment (matdenkas) [inline src/Frontend/ChartJson/Magnolia_Bay.json:1]: Cant review, ive no clue what this is lol

File: src/tests/IntegrationTests/test_HOHONU.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
```
Comment (matdenkas) [inline src/tests/IntegrationTests/test_HOHONU.py:1]: THESE ARE WAAAAAAY BETTER tests! Thank you for making that change. However im afraid they are failing on my machine.

image

File: tools/DatabaseMigration/2_4/2_4_DatabaseMigration.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
```
Comment (matdenkas) [inline tools/DatabaseMigration/2_4/2_4_DatabaseMigration.py:1]: Perfect! Works going up and rolling back!

File: src/DataIngestion/DI_Classes/HOHONU.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    def __convert_dataframe_to_input(self, df: pd.DataFrame) ->list[Output]:
+        """Converts a dataframe 
+        :param date: A date time
+        """
+        df = df.reset_index()
+        inputs = []
+        for timestamp, value in zip(df["timestamp"], df["value"]):
+            dt = timestamp.to_pydatetime()
+            dataPoint = Input(
+                dataValue= value,
```
Comment (matdenkas) [inline src/DataIngestion/DI_Classes/HOHONU.py:123]: TY for fixing this!

**Review by matdenkas — CHANGES_REQUESTED** [review]
This is a tone of work and it looks very well done. Tests are currently failing on my machine though after building. Once we know they are passing then this LGTM!!!!

**Review by lovelysandlonelys — CHANGES_REQUESTED** [review]
The tests are also failing for me! I think that it might be some kind of importing issue 😥 

image

This is a screenshot of some testing I ran. All of the results.data items ARE Inputs, however isinstance() is treating 'Input' and '< DataClasses.Input' > as two differnt things. Maybe! I just poked around a little. 

here's the code i used: 
for item in result.data :
            print(f'input: {item} \n[Type] -> {type(item)}')
            print(f'Is Instance of Input: {isinstance(item, Input)}')
            print(f'IDs: ')
            print(f'Input ID {id(Input)}')
            print(f'item ID {id(type(item))}')

File: src/tests/IntegrationTests/test_HOHONU.py (hunk 4 above)
Comment (matdenkas) [reply src/tests/IntegrationTests/test_HOHONU.py:1]: All tests passing

File: src/tests/IntegrationTests/test_HOHONU.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+def test_ingest_series(seriesDescription: SeriesDescription, timeDescription: TimeDescription, expected_output: None):
+    """This function tests the test ingest series method. This test makes sure that the HOHONU ingestion class is being properly mapped
+        function is complete.
+    """
+    load_dotenv()
+    hohonu = data_ingestion_factory(seriesDescription)
+    result = hohonu.ingest_series(seriesDescription, timeDescription)
+   
+    if result is None:
+        assert False, "The result is None"
```
Comment (matdenkas) [inline src/tests/IntegrationTests/test_HOHONU.py:37]: I had no idea you could just tack a string onto an assertion! Rad

**Review by matdenkas — APPROVED** [review]
LGTM! That was a slog! You deserve ice cream!

**Pull request merged**